libimage/load.go

```diff
@@ -163,13 +187,34 @@ func (r *Runtime) loadMultiImageDockerArchive(ctx context.Context, ref types.Ima
 	var copiedImages []string
 	for _, list := range refLists {
 		for _, listRef := range list {
-			names, err := r.copyFromDockerArchiveReaderReference(ctx, reader, listRef, options)
+			images, err := r.copyFromDockerArchiveReaderReference(ctx, reader, listRef, options)
 			if err != nil {
 				return nil, err
 			}
-			copiedImages = append(copiedImages, names...)
+			names := []string{}
+			for _, img := range images {
+				if len(img.Names()) > 0 {
+					names = append(names, img.Names()...)
+				} else {
+					// If no name is specified for this image
+					// then return its ID()
+					names = append(names, "sha256:"+img.ID())
+				}
+			}
+			for _, name := range names {
+				copiedImages = appendIfNotPresent(copiedImages, name)
+			}
 		}
 	}
 
 	return copiedImages, nil
 }
+
+func appendIfNotPresent(slice []string, element string) []string {
+	for _, v := range slice {
```

Comment:
`slices.Contains`. (I suppose it isn’t worth it to build an ordered deduplicated list with amortized O(1) append…)

---

Non-blocking: Ultimately I think this is a bad _abstraction_: The caller already loops over `names` to collect the values, and _then_ to call this function. The caller could, instead, maintain a `set` (not in the language :/ ) of all known values, and add only the new ones, directly. Or, the caller could use an ordered deduplicated list data structure.

Splitting the caller’s “create `copiedImages`” loop in _this_ place makes it invisible how the caller + callee combine into an O(N^2) operation.

… but then again, the typical length of the array is probably 2.